Use intra-doc links for `core/src/slice.mod.rs`

Amjad50 opened the pull request:
partial help in #75080 

r? @jyn514

- most are using primitive types links, which cannot be used with intra links at the moment
- also `std` cannot be referenced in any link, `std::ptr::NonNull` and `std::slice` could not be referenced

Amjad50 committed:
Use intra-doc links for `core/src/slice.mod.rs`

- most are using primitive types links, which cannot be used with intra
  links at the moment
- also `std` cannot be referenced in any link

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @jyn514 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

jyn514 commented on library/core/src/slice/mod.rs:
```diff
@@ -5390,7 +5390,7 @@ unsafe impl<'a, T> TrustedRandomAccess for ChunksMut<'a, T> {
 ///
 /// [`chunks_exact`]: ../../std/primitive.slice.html#method.chunks_exact
-/// [`remainder`]: ../../std/slice/struct.ChunksExact.html#method.remainder
+/// [`remainder`]: crate::slice::ChunksExact::remainder
```
This is in the `slice` module, so `ChunksExact` is already in scope.
```suggestion
/// [`remainder`]: ChunksExact::remainder
```

jyn514 commented on library/core/src/slice/mod.rs:
```diff
@@ -5530,7 +5530,7 @@ unsafe impl<'a, T> TrustedRandomAccess for ChunksExact<'a, T> {
 ///
 /// [`chunks_exact_mut`]: ../../std/primitive.slice.html#method.chunks_exact_mut
-/// [`into_remainder`]: ../../std/slice/struct.ChunksExactMut.html#method.into_remainder
+/// [`into_remainder`]: crate::slice::ChunksExactMut::into_remainder
```
```suggestion
/// [`into_remainder`]: ChunksExactMut::into_remainder
```

jyn514 commented on library/core/src/slice/mod.rs:
```diff
@@ -5390,7 +5390,7 @@ unsafe impl<'a, T> TrustedRandomAccess for ChunksMut<'a, T> {
 /// This struct is created by the [`chunks_exact`] method on [slices].
 ///
 /// [`chunks_exact`]: ../../std/primitive.slice.html#method.chunks_exact
```
This is #63351 , no need to change anything.

Amjad50 replied:
this applies to all primitives right? like this:
```
/// [slices]: ../../std/primitive.slice.html
```

jyn514 replied:
Not all primitives, but to all methods on slices. See the issue for a full list; it's mostly slices and pointers.

jyn514 commented on library/core/src/slice/mod.rs:
```diff
@@ -5667,7 +5667,7 @@ unsafe impl<'a, T> TrustedRandomAccess for ChunksExactMut<'a, T> {
 ///
 /// [`array_chunks`]: ../../std/primitive.slice.html#method.array_chunks
-/// [`remainder`]: ../../std/slice/struct.ArrayChunks.html#method.remainder
+/// [`remainder`]: crate::slice::ArrayChunks::remainder
```
```suggestion
/// [`remainder`]: ArrayChunks::remainder
```

jyn514 commented on library/core/src/slice/mod.rs:
```diff
@@ -6470,7 +6470,7 @@ pub unsafe fn from_raw_parts<'a, T>(data: *const T, len: usize) -> &'a [T] {
 /// [`NonNull::dangling()`]: ../../std/ptr/struct.NonNull.html#method.dangling
 /// [`pointer::offset`]: ../../std/primitive.pointer.html#method.offset
-/// [`from_raw_parts`]: ../../std/slice/fn.from_raw_parts.html
+/// [`from_raw_parts`]: crate::slice::from_raw_parts
```
`from_raw_parts` is in scope, so it will be linked to automatically.
```suggestion
```

jyn514 commented on library/core/src/slice/mod.rs:
```diff
@@ -6072,7 +6072,7 @@ unsafe impl<'a, T> TrustedRandomAccess for RChunksMut<'a, T> {
 ///
 /// [`rchunks_exact`]: ../../std/primitive.slice.html#method.rchunks_exact
-/// [`remainder`]: ../../std/slice/struct.ChunksExact.html#method.remainder
+/// [`remainder`]: crate::slice::ChunksExact::remainder
```
```suggestion
/// [`remainder`]: ChunksExact::remainder
```

jyn514 commented on library/core/src/slice/mod.rs:
```diff
@@ -6217,7 +6217,7 @@ unsafe impl<'a, T> TrustedRandomAccess for RChunksExact<'a, T> {
 ///
 /// [`rchunks_exact_mut`]: ../../std/primitive.slice.html#method.rchunks_exact_mut
-/// [`into_remainder`]: ../../std/slice/struct.ChunksExactMut.html#method.into_remainder
+/// [`into_remainder`]: crate::slice::ChunksExactMut::into_remainder
```
```suggestion
/// [`into_remainder`]: ChunksExactMut::into_remainder
```

jyn514 reviewed (commented):
There are a lot more links that can still be converted (e.g. https://github.com/rust-lang/rust/blob/197187f1f176870007700d55a4dfef646de3343f/library/core/src/slice/mod.rs#L7), but this is a good partial fix :)

r=me with comments addressed

jyn514 commented:
> also std cannot be referenced in any link

Use `crate` for items that are in `core` and re-exported in std. If there items are in std or alloc only, you can't use intra-doc links for them.

Amjad50 commented:
> There are a lot more links that can still be converted (e.g.
> 
> https://github.com/rust-lang/rust/blob/197187f1f176870007700d55a4dfef646de3343f/library/core/src/slice/mod.rs#L7
> 
> ), but this is a good partial fix :)
> r=me with comments addressed

Thanks so much for the coverage :+1: .

I'm not sure how can I fix this, I don't think I can use `crate::slice` right?

jyn514 commented:
You can :) This is linking to https://doc.rust-lang.org/std/slice/index.html, which is a re-export of https://doc.rust-lang.org/core/slice/index.html, which is the current module. So `crate::slice` or `self` will go to the same place.

jyn514 commented:
Oh hmm - it looks like that's not quite true, the `std` page has more info than `core`. So you can't change that link, sorry.

Amjad50 commented:
Yah, thats what made me think of keeping it.

Amjad50 committed:
Fixes `intra-doc` same scope links

Amjad50 commented:
Even though this is 7k file, but not many changes

jyn514 commented:
You could also link to `core::ptr` and `core::ptr::NonNull`: https://github.com/rust-lang/rust/blob/042d6902d26d283d373867cb7997ba463be51ce3/library/core/src/slice/mod.rs#L6430-L6431

Amjad50 commented:
Ok, can I link
```
 /// [valid]: ../../std/ptr/index.html#safety
```
?

Amjad50 committed:
Used intra-doc links for  `NonNull::dangling()` occurrences

Amjad50 commented:
It appears yes, I can

Amjad50 committed:
Used intra-doc links for `ptr#safety` occurrences

jyn514 commented:
```
alloc/slice/fn.from_raw_parts.html:18: broken link - alloc/slice/crate::ptr::NonNull::dangling
alloc/slice/fn.from_raw_parts_mut.html:17: broken link - alloc/slice/crate::ptr::NonNull::dangling
```
This is https://github.com/rust-lang/rust/issues/65983, fixed by https://github.com/rust-lang/rust/pull/73101. If you wait a day or two until the beta bump it will be fixed then; if not you'll have to revert the changes.

Amjad50 commented:
sorry I didn't run the checker for that.

Ok I guess we can wait for it to be fixed.

Thanks again for the help

jyn514 commented:
> sorry I didn't run the checker for that.

That's ok, that's what CI is for :)

jyn514 commented:
Beta was bumped today, so this should be good to go after a rebase :)

Amjad50 commented:
still broken :(

jyn514 commented:
@amjad50 you haven't rebased, so it's showing the test result from the previous compiler. Try rebasing over master.

Amjad50 commented:
I rebased, when I push to here should I force?

jyn514 commented:
You should force-push, yes. `git push -f`

jyn514 commented:
If you merge it defeats the point of rebasing: it shows both the old and new commits, and makes it look in the diff as if you have all the hundreds of commits that you rebased over.

Amjad50 commented:
Is the rebase done correctly? It failed

jyn514 commented:
Your rebase is fine but the bug was still not fixed :( It's a rustdoc bug of some sort, I'll try to look into it this weekend.

If you can reproduce outside the standard library that would be amazing ❤️ but no problem if you're not sure how.

Amjad50 commented:
Definitely, I really love contributing. Thanks so much for the help.😊

jyn514 commented:
I don't understand why this is failing :/ the error is being returned directly from `rustc_resolve`:

```
$ RUSTDOC_LOG=rustdoc::passes::collect=debug,rustc_resolve=debug xpy test --stage 1 linkchecker library/alloc
2020-08-29T22:39:13.076771Z DEBUG rustc_resolve: resolve_path(path=[Segment { ident: 
crate#0, id: Some(NodeId(101628)), has_generic_args: false }], opt_ns=Some(TypeNS), r
ecord_used=true, path_span=library/alloc/src/lib.rs:1:1: 1:1 (#0), crate_lint=No)
2020-08-29T22:39:13.076778Z DEBUG rustc_resolve: resolve_path ident 0 crate#0 Some(No
deId(101628))
2020-08-29T22:39:13.076780Z DEBUG rustc_resolve: resolve_crate_root(crate#0)
2020-08-29T22:39:13.076782Z DEBUG rustc_resolve: resolve_crate_root: not DollarCrate
2020-08-29T22:39:13.076800Z DEBUG rustc_resolve: resolve_crate_root(crate#0): found n
o mark (ident.span = library/alloc/src/lib.rs:1:1: 1:1 (#0))
2020-08-29T22:39:13.076805Z DEBUG rustc_resolve: resolve_path(path=[Segment { ident: 
crate#0, id: Some(NodeId(101629)), has_generic_args: false }, Segment { ident: ptr#0,
 id: Some(NodeId(101630)), has_generic_args: false }], opt_ns=Some(ValueNS), record_u
sed=true, path_span=library/alloc/src/lib.rs:1:1: 1:1 (#0), crate_lint=No)
2020-08-29T22:39:13.076810Z DEBUG rustc_resolve: resolve_path ident 0 crate#0 Some(No
deId(101629))
2020-08-29T22:39:13.076812Z DEBUG rustc_resolve: resolve_crate_root(crate#0)
2020-08-29T22:39:13.076814Z DEBUG rustc_resolve: resolve_crate_root: not DollarCrate
2020-08-29T22:39:13.076817Z DEBUG rustc_resolve: resolve_crate_root(crate#0): found n
o mark (ident.span = library/alloc/src/lib.rs:1:1: 1:1 (#0))
2020-08-29T22:39:13.076820Z DEBUG rustc_resolve: resolve_path ident 1 ptr#0 Some(Node
Id(101630))
2020-08-29T22:39:13.077589Z DEBUG rustdoc::passes::collect_intra_doc_links: crate::pt
r resolved to Err(()) in namespace ValueNS with parent DefId(1:6606 ~ core[c85e]::sli
ce[0])
2020-08-29T22:39:13.077601Z DEBUG rustc_resolve: resolve_path(path=[Segment { ident: crate#0, id: Some(NodeId(101631)), has_generic_args: false }], opt_ns=Some(TypeNS), record_used=true, path_span=library/alloc/src/lib.rs:1:1: 1:1 (#0), crate_lint=No)
2020-08-29T22:39:13.077608Z DEBUG rustc_resolve: resolve_path ident 0 crate#0 Some(NodeId(101631))
2020-08-29T22:39:13.077611Z DEBUG rustc_resolve: resolve_crate_root(crate#0)
2020-08-29T22:39:13.077613Z DEBUG rustc_resolve: resolve_crate_root: not DollarCrate
2020-08-29T22:39:13.077616Z DEBUG rustc_resolve: resolve_crate_root(crate#0): found no mark (ident.span = library/alloc/src/lib.rs:1:1: 1:1 (#0))
2020-08-29T22:39:13.077638Z  INFO rustdoc::passes::collect_intra_doc_links: ignoring warning from parent crate: unresolved link to `crate::ptr`
```

Amjad50 committed:
Fix intra-doc path resolution problem in `library/alloc/src/slice.rs`

`alloc::slice` uses `core::slice` functions, documentation are copied
from there and the links as well without resolution. `crate::ptr...`
cannot be resolved in `alloc::slice`, but `ptr` itself is imported in
both `alloc::slice` and `core::slice`, so we used that instead.

Amjad50 commented:
I think I fixed it, the problem in general is not solved. But for this case, we can use `ptr` instead of `crate::ptr` as it is available in both `alloc::slice` and `core::slice`

Amjad50 commented:
So from what I understand, `use` will copy the function/struct into that location together with it's documentation, and then resolve it from there.

Is there a way to change it so it resolves first and then copy the documentation? And would that affect other documentation processes.

jyn514 commented:
> So from what I understand, `use` will copy the function/struct into that location together with it's documentation, and then resolve it from there.
> 
> Is there a way to change it so it resolves first and then copy the documentation? And would that affect other documentation processes.

The fact that it does this is a bug, and I thought I fixed it in https://github.com/rust-lang/rust/pull/73101. I don't know why it's still happening.

jyn514 commented:
In fact if you look at the debug logging it _shouldn't_ be resolving in the current scope: it says itself that it's looking at core: `parent DefId(1:6606 ~ core[c85e]::slice[0])`

Amjad50 commented:
Should this be merged? Or you think it's better to fix the problem first?

jyn514 commented:
Let's merge this for now, since the link will keep working even if the rustdoc bug is fixed. I'll open a separate issue for the rustdoc bug.

@bors r+ rollup

bors commented:
:pushpin: Commit 300a0072a27f9c2d64ea63cea3b15463383683b0 has been approved by `jyn514`

<!-- @bors r=jyn514 300a0072a27f9c2d64ea63cea3b15463383683b0 -->
<!-- homu: {"type":"Approved","sha":"300a0072a27f9c2d64ea63cea3b15463383683b0","approver":"jyn514"} -->

bors commented:
:hourglass: Testing commit 300a0072a27f9c2d64ea63cea3b15463383683b0 with merge 8bfe289886af727150c1b9ec502cbfd7bbf425e8...
<!-- homu: {"type":"BuildStarted","head_sha":"300a0072a27f9c2d64ea63cea3b15463383683b0","merge_sha":"8bfe289886af727150c1b9ec502cbfd7bbf425e8"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1050632318), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36327)
Approved by: jyn514
Pushing 8bfe289886af727150c1b9ec502cbfd7bbf425e8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jyn514","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36327","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1050632318"},"merge_sha":"8bfe289886af727150c1b9ec502cbfd7bbf425e8"} -->
